k2/csrc/tensor_ops.cu

```diff
@@ -507,4 +507,249 @@ Tensor SimpleRaggedIndexSelect1D(Tensor &src, Ragged<int32_t> &indexes) {
   return ans;
 }
 
+template <typename Real>
+struct DiscountedCumSumElement {
+  Real y;      // y is the partial sums of x values.  Initially it is just a
+               // single x value.  In general each x is multiplied by all
+               // previous gamms.
+  Real gamma;  // gamma is the product of gammas along a range of elements
+};
+template <typename Real>
+struct CombineCumSumOp {
+  __device__ DiscountedCumSumElement<Real> operator() (
+      DiscountedCumSumElement<Real> &a,
+      DiscountedCumSumElement<Real> &b) const {
+    return DiscountedCumSumElement<Real>{b.y +  b.gamma * a.y, a.gamma * b.gamma};
+  }
+};
+
+// A stateful callback functor that maintains a running prefix to be applied
+// during consecutive scan operations.
+template <typename Real>
+struct BlockPrefixCallbackOp
+{
+  using Elem = DiscountedCumSumElement<Real>;
+  Elem running_total;
+  // Constructor
+  __device__ BlockPrefixCallbackOp(): running_total{0.0, 0.0} { }
+  // Callback operator to be entered by the first warp of threads in the block.
+  // Thread-0 is responsible for returning a value for seeding the block-wide scan.
+  __device__ Elem operator()(Elem block_aggregate)
+    {
+      Elem old_prefix = running_total;
+      running_total = block_aggregate;
+      return old_prefix;
+    }
+};
+
+/*
+  Notes for DiscountedCumSum.
+
+    It implements a discounted sum along a sequence.  Suppose we have x_i, gamma_i and
+  y_i, for 0 <= i < T.  Then we do:
+      y_0 = x_0
+      y_i = x_i + y_{i-1} gamma_i
+   for 0 < i < T.  (This is done as a generic inclusive-scan/inclusive-sum with a special
+   reduction op).
+
+  See DiscountedCumSumElement and CombineCumSumOp for how we use a special operator to
+  do this as an inclusive-sum.
+
+  The tensors involved must be 2-dimensional with dimensions (N, T) where N is
+  the batch size and T the time duration, and the stride on the T dim must be 1.
```

Comment:
> the stride on the T dim must be 1.

I think this constraint no longer exists.

Also, the documentation about the `stride1` for the arguments `gamma` and `y` should
also be updated.